In [1]:
import pandas as pd
import numpy as np
import pulp


# Add kaliya lincoln

In [2]:
blakely = pd.read_csv('../Cleaning/Individuals_Results/blakely.csv')
biles = pd.read_csv('../Cleaning/Individuals_Results/biles.csv')
carey = pd.read_csv('../Cleaning/Individuals_Results/carey.csv')
chiles = pd.read_csv('../Cleaning/Individuals_Results/chiles.csv')
dicello = pd.read_csv('../Cleaning/Individuals_Results/dicello.csv')
jones = pd.read_csv('../Cleaning/Individuals_Results/jones.csv')
jong = pd.read_csv('../Cleaning/Individuals_Results/jong.csv')
lee = pd.read_csv('../Cleaning/Individuals_Results/lee.csv')
matthews = pd.read_csv('../Cleaning/Individuals_Results/matthews.csv')
miller = pd.read_csv('../Cleaning/Individuals_Results/miller.csv')
roberson = pd.read_csv('../Cleaning/Individuals_Results/roberson.csv')
sullivan = pd.read_csv('../Cleaning/Individuals_Results/sullivan.csv')
suman = pd.read_csv('../Cleaning/Individuals_Results/suman.csv')
wong = pd.read_csv('../Cleaning/Individuals_Results/wong.csv')
lincoln = pd.read_csv('../Cleaning/Individuals_Results/lincoln.csv')


In [3]:
carey.head(1)

Name        Date    Competition Round      VT   UB   BB      FX   AA
0  Jade Carey  2021-07-23  Olympic Games    EF  12.416  0.0  0.0  14.366  0.0

In [4]:
chiles.head(1)

Name        Date    Competition Round      VT      UB      BB  \
0  Jordan Chiles  2021-07-23  Olympic Games    TF  14.666  14.166  13.433   

     FX      AA  
0  11.7  53.965

In [5]:
lincoln.head(1)

Name  year         Competition Round   VT    UB    BB   FX   AA
0  Kaliya Lincoln  2023  U.S. Worlds Trials    EF  0.0  12.6  13.1  0.0  0.0

In [6]:
dicello.head(1)

Name        Date          Competition Round   VT   UB      BB  \
0  Kayla DiCello  2021-10-18  World Championships    EF  0.0  0.0  11.866   

       FX   AA  
0  13.633  0.0

In [7]:
jones.head(1)

Name        Date          Competition Round    VT    UB    BB  \
0  Shilese Jones  2021-06-24  U.S. Olympic Trials    AA  14.9  12.9  13.3   

       FX      AA  
0  13.533  54.633

In [8]:
jong.head(1)

Name        Date                Competition Round      VT   UB  \
0  Katelyn Jong  2021-11-25  Junior Pan American Games    EF  13.933  0.0   

     BB    FX   AA  
0  12.7  13.0  0.0

In [9]:
biles = biles[biles['AA'] != 0]
blakely = blakely[blakely['AA'] != 0]
carey = carey[carey['AA'] != 0]
chiles = chiles[chiles['AA'] != 0]
dicello = dicello[dicello['AA'] != 0]
jones = jones[jones['AA'] != 0]
jong = jong[jong['AA'] != 0]
lee = lee[lee['AA'] != 0]
matthews = matthews[matthews['AA'] != 0]
miller = miller[miller['AA'] != 0]
roberson = roberson[roberson['AA'] != 0]
sullivan = sullivan[sullivan['AA'] != 0]
suman = suman[suman['AA'] != 0]
wong = wong[wong['AA'] != 0]
lincoln = lincoln[lincoln['AA'] != 0]

In [10]:
biles.rename(columns={'Gymnast': 'Name'}, inplace=True)

In [12]:
df1 = pd.concat([blakely, biles, carey, chiles, dicello,jones,jong,lee,matthews, miller,
                        roberson,sullivan, suman,wong, ], axis=0)

In [13]:
df1 = df1[df1['AA'] != 0]

In [14]:
df1['year'] = df1['Date'].str[:4]

In [15]:
df1=df1.drop(columns=['Date', 'Competition', 'Round'])

In [16]:
lincoln=lincoln.drop(columns=['Competition', 'Round'])

In [17]:
df = pd.concat([df1, lincoln ], axis=0)

In [19]:
df.info()

<class 'pandas.core.frame.DataFrame'>
Index: 225 entries, 0 to 7
Data columns (total 7 columns):
 #   Column  Non-Null Count  Dtype  
---  ------  --------------  -----  
 0   Name    225 non-null    object 
 1   VT      225 non-null    object 
 2   UB      225 non-null    float64
 3   BB      225 non-null    float64
 4   FX      225 non-null    float64
 5   AA      225 non-null    float64
 6   year    225 non-null    object 
dtypes: float64(4), object(3)
memory usage: 14.1+ KB


In [20]:

df['VT'] = df['VT'].astype(float)

In [21]:
df['year'].value_counts()

year
2021    89
2022    71
2023    60
2023     5
Name: count, dtype: int64

In [22]:
df['year'] = df['year'].str.strip()

In [23]:
df['year'].fillna(2022, inplace=True)

In [24]:
df['year'] = df['year'].astype(int)

In [25]:
dsfa

NameError: name 'dsfa' is not defined

In [26]:
df.info()

<class 'pandas.core.frame.DataFrame'>
Index: 225 entries, 0 to 7
Data columns (total 7 columns):
 #   Column  Non-Null Count  Dtype  
---  ------  --------------  -----  
 0   Name    225 non-null    object 
 1   VT      225 non-null    float64
 2   UB      225 non-null    float64
 3   BB      225 non-null    float64
 4   FX      225 non-null    float64
 5   AA      225 non-null    float64
 6   year    225 non-null    int64  
dtypes: float64(5), int64(1), object(1)
memory usage: 14.1+ KB


In [27]:
df=df.drop(columns=['year'])

In [28]:
grouped_means = df.groupby('Name').mean()

In [29]:
grouped_means

VT         UB         BB         FX         AA
Name                                                                     
Ashlee Sullivan     13.477778  12.427778  12.888889  13.077778  51.872222
Jade Carey          14.615308  13.271692  12.763923  13.687000  54.337923
Jordan Chiles       14.630062  14.089438  13.383188  13.580125  55.682812
Joscelyn Roberson   13.892105  11.328947  12.926316  13.118421  51.265789
Kaliya Lincoln      13.860000  11.990000  12.780000  13.990000  52.620000
Katelyn Jong        13.950000  13.137500  12.350000  13.105000  52.542500
Kayla DiCello       14.187500  13.423875  13.018688  13.720813  54.350875
Leanne Wong         14.353563  13.385312  13.300000  13.596813  54.635688
Nola Matthews       13.054348  13.202174  12.650000  12.570304  51.476826
Shilese Jones       14.511067  14.248733  13.026667  13.545533  55.332000
Simone Biles        15.348200  14.224900  14.544900  14.749900  58.867900
Skye Blakely        14.119000  13.648786  13.444071  13.213071  54.424929
Sunisa Lee          14.334583  14.877750  14.109583  13.877667  57.199583
Tiana Sumanasekera  13.958333  11.318500  13.312222  13.107444  51.698167
Zoe Miller          13.642105  13.856947  12.322737  12.335053  52.156842

In [30]:
# unique_names = df['Name'].unique()

# # Create a dictionary of DataFrames, one for each unique 'Name'
# dataframes = {}
# for name in unique_names:
#     dataframes[name] = df[df['Name'] == name]


In [31]:
# grouped_dataframes = dict(iter(df.groupby('Name')))

In [32]:
# # Group the DataFrame by 'Name' and store in a dictionary
# grouped_dataframes = dict(iter(df.groupby('Name')))

# # Access and use individual DataFrames by their 'Name'
# for name, sub_df in grouped_dataframes.items():
#     # You can access and manipulate each DataFrame by its name
#     print(f"Data for {name}:")
#     print(sub_df)
#     # You can perform other operations on each DataFrame as needed
#     # For example, you can save them to separate CSV files:
#     sub_df.to_csv(f"{name}_data.csv", index=False)
#     print()


In [38]:
# Sample athlete scores (replace with the provided data)
athlete_scores = {
    'Ashlee Sullivan': {'VT': 13.477778, 'UB': 12.427778, 'BB': 12.888889, 'FX': 13.077778, 'AA': 51.872222},
    'Jade Carey': {'VT': 14.615308, 'UB': 13.271692, 'BB': 12.763923, 'FX': 13.687000, 'AA': 54.337923},
    'Jordan Chiles': {'VT': 14.630062, 'UB': 14.089438, 'BB': 13.383188, 'FX': 13.580125, 'AA': 55.682812},
    'Joscelyn Roberson': {'VT': 13.892105, 'UB': 11.328947, 'BB': 12.926316, 'FX': 13.118421, 'AA': 51.265789},
    'Katelyn Jong': {'VT': 13.950000, 'UB': 13.137500, 'BB': 12.350000, 'FX': 13.105000, 'AA': 52.542500},
    'Kayla DiCello': {'VT': 14.187500, 'UB': 13.423875, 'BB': 13.018688, 'FX': 13.720813, 'AA': 54.350875},
    'Leanne Wong': {'VT': 14.353563, 'UB': 13.385312, 'BB': 13.300000, 'FX': 13.596813, 'AA': 54.635688},
    'Nola Matthews': {'VT': 13.054348, 'UB': 13.202174, 'BB': 12.650000, 'FX': 12.570304, 'AA': 51.476826},
    'Shilese Jones': {'VT': 14.511067, 'UB': 14.248733, 'BB': 13.026667, 'FX': 13.545533, 'AA': 55.332000},
    'Simone Biles': {'VT': 15.348200, 'UB': 14.224900, 'BB': 14.544900, 'FX': 14.749900, 'AA': 58.867900},
    'Skye Blakely': {'VT': 14.119000, 'UB': 13.648786, 'BB': 13.444071, 'FX': 13.213071, 'AA': 54.424929},
    'Sunisa Lee': {'VT': 14.334583, 'UB': 14.877750, 'BB': 14.109583, 'FX': 13.877667, 'AA': 57.199583},
    'Tiana Sumanasekera': {'VT': 13.958333, 'UB': 11.318500, 'BB': 13.312222, 'FX': 13.107444, 'AA': 51.698167},
    'Zoe Miller': {'VT': 13.642105, 'UB': 13.856947, 'BB': 12.322737, 'FX': 12.335053, 'AA': 52.156842},
    'Kaliya Lincoln': {'VT': 13.860000, 'UB': 11.990000, 'BB': 12.780000, 'FX': 13.990000, 'AA': 52.620000},
}


# Create a LP problem
problem = pulp.LpProblem("Athlete_Selection", pulp.LpMaximize)

# Define decision variables for each athlete
athletes = list(athlete_scores.keys())
athlete_vars = pulp.LpVariable.dicts("Select", athletes, cat=pulp.LpBinary)

# Define the objective to maximize total scores in the qualifying round
# You should replace 'Apparatus1' with the appropriate apparatus (e.g., 'VT', 'UB', 'BB', 'FX')
apparatus_to_maximize = 'VT'  # Change this apparatus as needed
problem += pulp.lpSum(athlete_scores[athlete][apparatus_to_maximize] * athlete_vars[athlete] for athlete in athletes)

# Constraints
problem += pulp.lpSum(athlete_vars[athlete] for athlete in athletes) == 5  # Select 5 athletes
problem += pulp.lpSum(athlete_vars[athlete] for athlete in athletes) >= 4  # Select at least 4 athletes

# Solve the LP problem
problem.solve()

# Display the selected athletes
selected_athletes = [athlete for athlete in athletes if pulp.value(athlete_vars[athlete]) == 1]
print("Selected Athletes:", selected_athletes)

Welcome to the CBC MILP Solver 
Version: 2.10.3 
Build Date: Dec 15 2019 

command line - /Users/ryantalbot/opt/anaconda3/envs/tf2/lib/python3.9/site-packages/pulp/solverdir/cbc/osx/64/cbc /var/folders/j_/555m2zps099832fjh_m8jjnc0000gn/T/ad0bc57d071a4190a9af115dff70ccf0-pulp.mps max timeMode elapsed branch printingOptions all solution /var/folders/j_/555m2zps099832fjh_m8jjnc0000gn/T/ad0bc57d071a4190a9af115dff70ccf0-pulp.sol (default strategy 1)
At line 2 NAME          MODEL
At line 3 ROWS
At line 7 COLUMNS
At line 83 RHS
At line 86 BOUNDS
At line 102 ENDATA
Problem MODEL has 2 rows, 15 columns and 30 elements
Coin0008I MODEL read with 0 errors
Option for timeMode changed from cpu to elapsed
Continuous objective value is 73.4582 - 0.00 seconds
Cgl0004I processed model has 1 rows, 15 columns (15 integer (15 of which binary)) and 15 elements
Cbc0038I Initial state - 0 integers unsatisfied sum - 0
Cbc0038I Solution found of -73.4582
Cbc0038I Before mini branch and bound, 15 integers at bou

<!-- The provided output is the result summary of a Mixed-Integer Linear Programming (MILP) optimization problem solved using the CBC (Coin-or Branch and Cut) solver. Here's a breakdown of the important parts of the output:

1. **CBC MILP Solver Information**:
   - This section provides information about the version and build date of the CBC MILP solver being used.

2. **Command Line**:
   - This line shows the command line used to run the CBC solver, specifying various options and the path to the solver executable.

3. **Problem Description**:
   - The lines starting with "At line X" describe the structure of the optimization problem, including rows (constraints), columns (variables), right-hand sides (RHS), bounds, and the end of the problem definition.

4. **Problem Statistics**:
   - "Problem MODEL has 2 rows, 14 columns, and 28 elements" provides information about the size of the optimization problem, including the number of rows (constraints) and columns (variables).

5. **Solution Information**:
   - "Continuous objective value" and "Integer solution" sections provide details about the objective value and the status of the solver:
     - "Continuous objective value" shows the objective value obtained in the continuous relaxation of the problem.
     - "Integer solution" indicates that an integer solution was found.

6. **Branch and Bound Algorithm**:
   - This section provides information about the progress of the branch and bound algorithm used to solve the MILP problem:
     - "Cbc0038I Initial state" indicates the initial state of the solver.
     - "Cbc0038I Solution found" reports the objective value found.
     - "Cbc0038I Mini branch and bound" indicates that no improvement was achieved in a mini branch and bound step.
     - "Cbc0035I Maximum depth" shows information about the depth and variables fixed.
     
7. **Cut Generation**:
   - This section reports the generation of cuts (inequalities) during the optimization process.

8. **Result**:
   - "Result - Optimal solution found" indicates that the solver found an optimal solution to the MILP problem.

9. **Objective Value and Solution**:
   - "Objective value" displays the final objective value obtained by the optimization.
   - "Selected Athletes" lists the names of the selected athletes as a result of the optimization.

10. **Solver Timing Information**:
    - This section provides timing information related to the solver's execution.

In summary, this output provides detailed information about how the CBC solver solved the MILP problem, including problem statistics, solver progress, and the final optimal solution. The "Selected Athletes" section shows the athletes chosen based on the optimization criteria. -->

In [39]:
selected_athletes

['Jade Carey', 'Jordan Chiles', 'Leanne Wong', 'Shilese Jones', 'Simone Biles']

---
The provided output is the result summary of a Mixed-Integer Linear Programming (MILP) optimization problem solved using the CBC (Coin-or Branch and Cut) solver. Here's a breakdown of the important parts of the output:

1. **CBC MILP Solver Information**:
   - This section provides information about the version and build date of the CBC MILP solver being used.

2. **Command Line**:
   - This line shows the command line used to run the CBC solver, specifying various options and the path to the solver executable.

3. **Problem Description**:
   - The lines starting with "At line X" describe the structure of the optimization problem, including rows (constraints), columns (variables), right-hand sides (RHS), bounds, and the end of the problem definition.

4. **Problem Statistics**:
   - "Problem MODEL has 2 rows, 14 columns, and 28 elements" provides information about the size of the optimization problem, including the number of rows (constraints) and columns (variables).

5. **Solution Information**:
   - "Continuous objective value" and "Integer solution" sections provide details about the objective value and the status of the solver:
     - "Continuous objective value" shows the objective value obtained in the continuous relaxation of the problem.
     - "Integer solution" indicates that an integer solution was found.

6. **Branch and Bound Algorithm**:
   - This section provides information about the progress of the branch and bound algorithm used to solve the MILP problem:
     - "Cbc0038I Initial state" indicates the initial state of the solver.
     - "Cbc0038I Solution found" reports the objective value found.
     - "Cbc0038I Mini branch and bound" indicates that no improvement was achieved in a mini branch and bound step.
     - "Cbc0035I Maximum depth" shows information about the depth and variables fixed.
     
7. **Cut Generation**:
   - This section reports the generation of cuts (inequalities) during the optimization process.

8. **Result**:
   - "Result - Optimal solution found" indicates that the solver found an optimal solution to the MILP problem.

9. **Objective Value and Solution**:
   - "Objective value" displays the final objective value obtained by the optimization.
   - "Selected Athletes" lists the names of the selected athletes as a result of the optimization.

10. **Solver Timing Information**:
    - This section provides timing information related to the solver's execution.

In summary, this output provides detailed information about how the CBC solver solved the MILP problem, including problem statistics, solver progress, and the final optimal solution. The "Selected Athletes" section shows the athletes chosen based on the optimization criteria.

The "Objective value" in the context of a Linear Programming (LP) or Mixed-Integer Linear Programming (MILP) problem represents the numerical value of the objective function at the optimal solution found by the solver. It's a critical piece of information as it indicates the value of the quantity that you are trying to either maximize or minimize in your optimization problem.

In the provided output:

    "Objective value: 73.45820000" means that the value of the objective function at the optimal solution found by the solver is approximately 73.4582.

Here's what this value represents:

    If your optimization problem is a maximization problem, this value tells you the maximum achievable value of the objective function under the given constraints. In this case, 73.4582 is the maximum possible value of the objective function.

    If your optimization problem is a minimization problem, this value tells you the minimum achievable value of the objective function. In this case, 73.4582 is the minimum possible value of the objective function.

In either case, this value is a key metric that helps you understand the quality of the solution. The objective value should align with your problem's goals. For example, if you're trying to maximize profits, a higher objective value is better. If you're trying to minimize costs, a lower objective value is better.

In practical terms, this objective value is often associated with the specific units or context of your problem. For example, in a financial optimization problem, it might represent the total profit in dollars. In a scheduling problem, it might represent the total completion time in hours.

So, in your context, an "Objective value" of 73.4582 indicates the quality of the solution based on the criteria defined in your MILP problem. The specific meaning of this value depends on the details of your optimization problem and what you are trying to achieve.